components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt

```diff
@@ -24,27 +27,117 @@ class DefaultLoginValidationDelegate(
+                savedLoginByGuid.username.isEmpty() -> {
+                    Operation.UPDATE
+                }
+                // When a user name changes, we create a new login
+                newLogin.username != savedLoginByGuid.username -> {
+                    Operation.CREATE
+                }
+                // Password has been updated
+                else -> {
+                    Operation.UPDATE
+                }
+            }
+        }
+    }
+
+    /**
+     * Verifies that the passed [Login] (merged with any known saved logins) is valid to be saved or updated.
+     * Not to be used for knowing whether to save or update.
+     */
     @Suppress("ComplexMethod") // This method is not actually complex
     override fun validateCanPersist(login: Login): Deferred<Result> {
         return scope.async {
+            // We check for a matching saved login by GUID and duplicate key elements
+            val savedLogin =
+                login.guid?.let { storage.get(it) } ?: getStoredLoginByElements(login)
+
+            // Check the merged login with the new password to display correct UI error states
+            val loginToCheck =
+                savedLogin?.mergeWithLogin(login)?.copy(password = login.password) ?: login
             try {
-                // We're setting guid=null here in order to trigger de-duping logic properly.
-                // Internally, the library ensures to not dupe records against themselves
-                // (via a `guid <> :guid` check), which means we need to "pretend" this is a new record,
-                // in order for `ensureValid` to actually throw `DUPLICATE_LOGIN`.
-                storage.ensureValid(login.copy(guid = null))
-                Result.CanBeCreated
+                storage.ensureValid(loginToCheck)
+                Result.Valid
             } catch (e: InvalidRecordException) {
                 when (e.reason) {
-                    InvalidLoginReason.DUPLICATE_LOGIN -> Result.CanBeUpdated
+                    InvalidLoginReason.DUPLICATE_LOGIN -> Result.Error.Duplicate
                     InvalidLoginReason.EMPTY_PASSWORD -> Result.Error.EmptyPassword
                     InvalidLoginReason.EMPTY_ORIGIN -> Result.Error.GeckoError(e)
                     InvalidLoginReason.BOTH_TARGETS -> Result.Error.GeckoError(e)
                     InvalidLoginReason.NO_TARGET -> Result.Error.GeckoError(e)
-                    // TODO in what ways can the login fields be illegal? represent these in the UI
                     InvalidLoginReason.ILLEGAL_FIELD_VALUE -> Result.Error.GeckoError(e)
                 }
             }
         }
     }
+
+    /**
+     * Searches storage to find a matching login to all A-S defined important elements
+     * (username, hostname, and formSubmitUrl or httpRealm)
+     * in case a user enters the same information as another login but doesn't match GUID
+     */
+    suspend fun getStoredLoginByElements(newLogin: Login): Login? {
```

Comment:
Nit: what about `getStoredLoginByImportantFields`?  Or `getStoredLoginByStructualEquality`?  Except it's only partial structural equality....   This one's hard to name.